base/src/H264DecoderNvCodecHelper.cpp

```diff
@@ -0,0 +1,754 @@
+	return true;
+}
+
+void NvDecoder::UnlockFrame(uint8_t** ppFrame, int nFrame)
+{
+	std::lock_guard<std::mutex> lock(m_mtxVPFrame);
+	m_vpFrame.insert(m_vpFrame.end(), &ppFrame[0], &ppFrame[nFrame]);
+}
+
+template<typename T>
+class YuvConverter {
+public:
+	YuvConverter(int nWidth, int nHeight) : nWidth(nWidth), nHeight(nHeight) {
+		pQuad = new T[nWidth * nHeight / 4];
+	}
+	~YuvConverter() {
+		delete pQuad;
+	}
+	void PlanarToUVInterleaved(T* pFrame, int nPitch = 0) {
+		if (nPitch == 0) {
+			nPitch = nWidth;
+		}
+		T* puv = pFrame + nPitch * nHeight;
+		if (nPitch == nWidth) {
+			memcpy(pQuad, puv, nWidth * nHeight / 4 * sizeof(T));
+		}
+		else {
+			for (int i = 0; i < nHeight / 2; i++) {
+				memcpy(pQuad + nWidth / 2 * i, puv + nPitch / 2 * i, nWidth / 2 * sizeof(T));
+			}
+		}
+		T* pv = puv + (nPitch / 2) * (nHeight / 2);
+		for (int y = 0; y < nHeight / 2; y++) {
+			for (int x = 0; x < nWidth / 2; x++) {
+				puv[y * nPitch + x * 2] = pQuad[y * nWidth / 2 + x];
+				puv[y * nPitch + x * 2 + 1] = pv[y * nPitch / 2 + x];
+			}
+		}
+	}
+	void UVInterleavedToPlanar(T* pFrame, int nPitch = 0) {
+		if (nPitch == 0) {
+			nPitch = nWidth;
+		}
+		T* puv = pFrame + nPitch * nHeight,
+			* pu = puv,
+			* pv = puv + nPitch * nHeight / 4;
+		for (int y = 0; y < nHeight / 2; y++) {
+			for (int x = 0; x < nWidth / 2; x++) {
+				pu[y * nPitch / 2 + x] = puv[y * nPitch + x * 2];
+				pQuad[y * nWidth / 2 + x] = puv[y * nPitch + x * 2 + 1];
+			}
+		}
+		if (nPitch == nWidth) {
+			memcpy(pv, pQuad, nWidth * nHeight / 4 * sizeof(T));
+		}
+		else {
+			for (int i = 0; i < nHeight / 2; i++) {
+				memcpy(pv + nPitch / 2 * i, pQuad + nWidth / 2 * i, nWidth / 2 * sizeof(T));
+			}
+		}
+	}
+
+private:
+	T* pQuad;
+	int nWidth, nHeight;
+};
+
+H264DecoderNvCodecHelper::H264DecoderNvCodecHelper(int mWidth, int mHeight)
+{
+	CUcontext cuContext;
+	bool bUseDeviceFrame = false;
+	cudaVideoCodec eCodec = cudaVideoCodec_H264;;
+	std::mutex* pMutex = NULL; 
+	bool bLowLatency = false; 
+	bool bDeviceFramePitched= false; int maxWidth; int maxHeight;
+	ck(cuInit(0));
+	int nGpu = 0;
+	int iGpu = 0;
+	ck(cuDeviceGetCount(&nGpu));
+	if (iGpu < 0 || iGpu >= nGpu) {
+		std::cout << "GPU ordinal out of range. Should be within [" << 0 << ", " << nGpu - 1 << "]" << std::endl;
+		return;
+	}
+	CUdevice cuDevice = 0;
+
+	ck(cuDeviceGet(&cuDevice, iGpu));
+	char szDeviceName[80];
+	ck(cuCtxCreate(&cuContext, 0, cuDevice));
+	helper.reset(new NvDecoder(cuContext, mWidth, mHeight, bUseDeviceFrame, eCodec, pMutex));
+}
+
+bool H264DecoderNvCodecHelper::init(std::function<void(frame_sp&)> _send, std::function<frame_sp()> _makeFrame)
+{
+	makeFrame = _makeFrame;
+	send = _send;
+	return false;
+}
+
+void H264DecoderNvCodecHelper::ConvertToPlanar(uint8_t* pHostFrame, int nWidth, int nHeight, int nBitDepth) {
+	if (nBitDepth == 8) {
+		// nv12->iyuv
+		YuvConverter<uint8_t> converter8(nWidth, nHeight);
+		converter8.UVInterleavedToPlanar(pHostFrame);
+	}
+	else {
```

Comment:
can this "else" handle all other values for "nBitDepth" ? it looks like it can only handle 16 
instead this should be an "else if", 
if(8) =>doSomething
else if(16) =>doSomethingElse
else raise error !

Please make sure we have a test case which tests the nBitDepth=16 example.

Reply:
Improved it to use if , elseif , else now.